storage/json/datasource.go

```diff
@@ -101,33 +114,50 @@ type RecordStream struct {
 	alias                         string
 	offset                        int
 	batchSize                     int
+
+	workerCtxCancel    func()
+	workerCloseErrChan chan error
 }
 
-func (rs *RecordStream) Close() error {
-	err := rs.file.Close()
-	if err != nil {
-		return errors.Wrap(err, "Couldn't close underlying file")
+func (rs *RecordStream) Close(ctx context.Context, storage storage.Storage) error {
+	rs.workerCtxCancel()
+	err := <-rs.workerCloseErrChan
+	if err == context.Canceled {
+	} else if err != nil {
+		return errors.Wrap(err, "couldn't stop json worker")
+	}
+
+	if err := rs.file.Close(); err != nil {
+		return errors.Wrap(err, "couldn't close underlying file")
+	}
+
+	if err := storage.DropAll(rs.streamID.AsPrefix()); err != nil {
+		return errors.Wrap(err, "couldn't clear storage with streamID prefix")
 	}
 
 	return nil
 }
 
-func (rs *RecordStream) RunWorker(ctx context.Context) {
+func (rs *RecordStream) RunWorker(ctx context.Context) error {
 	for { // outer for is loading offset value and moving file iterator
+		select {
+		case <-ctx.Done():
+			return nil
```

Comment:
Usually you'd just return the error here ctx.Err(), and just handle the Canceled and DeadlineExceeded errors above.